auth/ante.go

```diff
@@ -0,0 +1,147 @@
+	utils "plasma-mvp-sidechain/utils"
+	"reflect"
+)
+
+// NewAnteHandler returns an AnteHandler that checks signatures,
+// and deducts fees from the first signer.
+func NewAnteHandler(utxoMapper types.UTXOMapper, txIndex *uint16, feeAmount *uint64) sdk.AnteHandler {
+	return func(
+		ctx sdk.Context, tx sdk.Tx,
+	) (_ sdk.Context, _ sdk.Result, abort bool) {
+
+		sigs := tx.GetSignatures()
+		if len(sigs) == 0 {
+			return ctx,
+				sdk.ErrUnauthorized("no signers").Result(),
+				true
+		}
+
+		msg := tx.GetMsg()
+
+		_, ok := tx.(db.BaseTx)
+		if !ok {
+			return ctx, sdk.ErrInternal("tx must be in form of BaseTx").Result(), true
+		}
+
+		// Assert that number of signatures is correct.
+		var signerAddrs = msg.GetSigners()
+
+		if len(sigs) != len(signerAddrs) {
+			return ctx,
+				sdk.ErrUnauthorized("wrong number of signers").Result(),
+				true
+		}
+
+		spendMsg, ok := msg.(db.SpendMsg)
+		if !ok {
+			return ctx, sdk.ErrInternal("Msg must be of type SpendMsg").Result(), true
+		}
+		signBytes := spendMsg.GetSignBytes()
+
+		//Check that inputs are valid and having valid signatures
+		position1 := types.Position{spendMsg.Blknum1, spendMsg.Txindex1, spendMsg.Oindex1, spendMsg.DepositNum1}
+		res := processSig(ctx, utxoMapper, position1, signerAddrs[0], sigs[0], signBytes)
+
+		if !res.IsOK() {
+			return ctx, res, true
+		}
+
+		posSignBytes := position1.GetSignBytes()
+
+		//Check that confirm signature is valid
+		res = processConfirmSig(ctx, utxoMapper, position1, spendMsg.ConfirmSigs1, posSignBytes)
+		if !res.IsOK() {
+			return ctx, res, true
+		}
+		//Verify validity of second input
+		if utils.ValidAddress(spendMsg.Owner2) {
+			position2 := types.Position{spendMsg.Blknum2, spendMsg.Txindex2, spendMsg.Oindex2, spendMsg.DepositNum2}
+			res = processSig(ctx, utxoMapper, position2, signerAddrs[1], sigs[1], signBytes)
+			if !res.IsOK() {
+				return ctx, res, true
+			}
+
+			posSignBytes = position2.GetSignBytes()
+
+			res = processConfirmSig(ctx, utxoMapper, position2, spendMsg.ConfirmSigs2, posSignBytes)
+			if !res.IsOK() {
+				return ctx, res, true
+			}
+		}
+
+		// If DeliverTx() update fee
+		if !ctx.IsCheckTx() {
+			//Increment amount of fees collected
+			(*feeAmount) += spendMsg.Fee
+		}
+
+		// TODO: tx tags (?)
+		return ctx, sdk.Result{}, false // continue...
+	}
+}
+
+func processSig(
+	ctx sdk.Context, um types.UTXOMapper,
+	position types.Position, addr crypto.Address, sig sdk.StdSignature, signBytes []byte) (
+	res sdk.Result) {
+	// Check UTXO is not nil.
+	utxo := um.GetUTXO(ctx, position)
+	if utxo == nil {
+		return sdk.ErrUnknownRequest("UTXO trying to be spent, does not exist").Result()
+	}
+
+	//Checks that utxo owner equals address in the spendmsg
+	if !reflect.DeepEqual(utxo.GetAddress().Bytes(), addr.Bytes()) {
+		return sdk.ErrUnauthorized("signer does not match utxo owner").Result()
+	}
+
+	hash := ethcrypto.Keccak256(signBytes)
+	pubKey1, err1 := ethcrypto.SigToPub(hash, sig.Signature.Bytes()[5:])
```

Comment:
Why apply `[5:]` to the bytes?

Reply:
Good question. 

The Bytes() function for crypto.SignatureSecp256K1 calls UnmarshalBinaryBare() as seen [here](https://github.com/tendermint/go-crypto/blob/master/signature.go#L64). 

When this occurs, the signature goes from 65 bytes to 70 bytes on my local test. The first five bytes are always the same regardless of the testing instance, on my local test they are [22 225 254 234 65]. According to the amino spec concrete structs should be prefixed with 4 bytes. I am still trying to figure out exactly why calling UnmarshalBinaryBare is adding those 5 bytes of padding to the front of the sig. I'd like to replace the "[5:]" with something more concrete based on the amino repo. 

Let me know if you figure anything out.

Reply:
As per the amino spec, The first 4 bytes after the signature is marshaled corresponds to the prefix bytes, which are outlined [here](https://github.com/tendermint/tendermint/blob/26f633ed48441f72895b710f0e87b7b6c6791066/docs/specification/new-spec/encoding.md#signaturesecp256k1) . 

The 5th byte corresponds to the length of the byte array. Signatures should always be 65 bytes in length and unless the amino spec changes, there should continue to be 4 bytes used for prefixing. 

 `[5:]` could be replaced with `[amino.PrefixBytesLen + 1:]` but I think the former is sufficient for now, and the dependency on the amino spec should be noted in the code.

Reply:
I see. Good find, and yes agreed on including a comment in the code about the prefix bytes.